contracts/ExecutionNodeVersionBeacon.cdc

```diff
@@ -0,0 +1,425 @@
+                return true
+            } else if self.major == other.major && self.minor == other.minor && self.patch < other.patch {
+                return true
+            } else {
+                return false
+            }
+        }
+
+        /// Returns true if Semver core is less than or
+        /// equal to passed Semver core and false otherwise
+        pub fun coreLessThanOrEqualTo(_ other: Semver): Bool {
+            return self.coreLessThan(other) || self.coreEqualTo(other)
+        }
+
+        /// Returns true if Semver is equal to passed
+        /// Semver core and false otherwise
+        pub fun coreEqualTo(_ other: Semver): Bool {
+            return self.major == other.major && self.minor == other.minor && self.patch == other.patch
+        }
+
+        /// Returns true if Semver is *exactly* equal to passed
+        /// Semver and false otherwise
+        pub fun strictEqualTo(_ other: Semver): Bool {
+            return self.coreEqualTo(other) && self.preRelease == other.preRelease
+        }
+
+    }
+
+    /// Event emitted any time a change is made to versionTable
+    pub event ExecutionNodeVersionTableAddition(height: UInt64, version: Semver)
+    pub event ExecutionNodeVersionTableDeletion(height: UInt64, version: Semver)
+    /// Event emitted any time the version update buffer period or variance is updated
+    pub event ExecutionNodeVersionUpdateBufferChanged(newVersionUpdateBuffer: UInt64)
+    pub event ExecutionNodeVersionUpdateBufferVarianceChanged(newVersionUpdateBufferVariance: UFix64)
+
+    /// Canonical storage path for ExecutionNodeVersionKeeper
+    pub let ExecutionNodeVersionKeeperStoragePath: StoragePath
+
+    /// Table dictating minimum version - {blockHeight: Semver}
+    /// Should be ordered by block height (index) which is
+    /// enforced by insertion/deletion
+    access(contract) let versionTable: {UInt64: Semver}
+
+    /// Number of blocks between block height when version boundary is defined and
+    /// the version boundary being defined. Nodes can expect that changes to the versionTable
+    /// they will have at least this long to respond to a version boundary
+    access(contract) var versionUpdateBuffer: UInt64
+    /// Set expectations regarding how much the versionUpdateBuffer can vary between version boundaries
+    access(contract) var versionUpdateBufferVariance: UFix64
+
+    /// Sorted Array containing future block heights where a version boundary is defined
+    /// We'll maintain this as easy mechanism for determining proximal version block boundaries
+    access(contract) var upcomingBlockBoundaries: [UInt64]
+
+    /// Sorted Array containing historical block heights where version boundaries were defined
+    access(contract) var archivedBlockBoundaries: [UInt64]
+
+    /// Admin interface that manages the Execution Node versioning
+    /// maintained in this contract
+    pub resource interface ExecutionNodeVersionAdmin {
```

Comment:
I'm curious why you made this interface. Do you expect your or other people to define other resources that implement this in the future?

Reply:
I don't foresee others (outside of those managing EN versions) making implementations. It's primarily an artifact from scoping the contract, but I thought it might be helpful in case of a need for future extensibility. The immediate need doesn't demand an interface, however.